### PR TITLE
Only trigger a rebuild when someone reloads the Play app 

### LINKED ISSUE #2550: Continuous build with Play reload builds unnecessarily when multiple changes are made
If we have the following:
- Run `gradle -t runPlay`
- Make changes to a Scala file. In the background, Gradle starts building.
- Make more changes to a Scala file. The build in the background is still going.
- Switch to the browser and refresh the page.  Since #2085 has been fixed, the refresh will block until all pending changes have been incorporated. 

In this scenario, at worst, we have to wait for nearly two builds for what a user may see as "one logical change".  We may want an opt-in behavior that would allow Gradle to wait until a "go ahead" signal has been made before kicking off a build while still relying on Gradle to detect pending changes.

### PR DESCRIPTION
**Pull request by big-guy**
Fixes #2550

### CODE REVIEW

File: subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java [hunk 1]
```diff
@@ -69,9 +79,11 @@ public FileSystemChangeWaiter createChangeWaiter(PendingChangesListener listener
         private final Collection<FileWatcherEvent> eventsBeforeListening = new ArrayList<FileWatcherEvent>();
         private final Lock eventDeliveryLock = new ReentrantLock();
 
-        private ChangeWaiter(FileWatcherFactory fileWatcherFactory, final PendingChangesListener pendingChangesListener, long quietPeriodMillis, BuildCancellationToken cancellationToken) {
+        private ChangeWaiter(FileWatcherFactory fileWatcherFactory, final PendingChangesListener pendingChangesListener, long quietPeriodMillis, boolean triggeredBuild, BuildCancellationToken cancellationToken, BuildGateToken buildGateToken) {
```
Comment (bigdaz) [inline subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java]: I don't like pushing `buildGateToken.waitForOpen()` into here. The simplest improvement would be to wrap this `ChangeWaiter` in a `BuildGatedChangeWaiter`.

But even better if we split the `FileSystemChangeWaiter` into 2 things: a basic `ContinuousBuildMonitor` API that has `wait(Runnable notifier)` and another that has the various filesystem specific methods. This change might be out of scope for the current story.

File: subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/continuous/PlayReloadWaitingIntegrationTest.groovy [hunk 2]
```diff
@@ -63,10 +65,12 @@ class PlayReloadWaitingIntegrationTest extends PlayReloadIntegrationTest {
         }
     }
 
-    def "wait for changes to be built when a request comes in during initial app startup and there are pending changes"() {
+    @Unroll
+    def "wait for changes to be built when a request comes in during initial app startup and there are pending changes and build is triggered=#triggered"() {
```
Comment (bigdaz) [inline subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/continuous/PlayReloadWaitingIntegrationTest.groovy]: s/triggered/gated/ ?

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        lock.lock();
+        try {
+            gatekeepers.add(gatekeeper);
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void open(GateKeeper gatekeeper) {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java]: Yep, I like this a lot better.

If we restrict the number of gate keepers you can have, we'll need to do something about having multiple PlayRun tasks in a single build.

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public void waitForOpen() {
+        lock.lock();
+        try {
+            if (!gatekeepers.isEmpty()) {
+                if (isClosed()) {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java]: good catch.

File: subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java [hunk 5]
```diff
@@ -33,31 +34,40 @@
 
 public class DefaultFileSystemChangeWaiterFactory implements FileSystemChangeWaiterFactory {
     public static final String QUIET_PERIOD_SYSPROP = "org.gradle.internal.filewatch.quietperiod";
+    public static final String TRIGGER_BUILD_SYSPROP = "org.gradle.internal.continuous.trigger";
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java]: It should be gated.

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java (hunk 3 above)
Comment (bigdaz) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildGateToken.java]: Is this a real use case? I guess we probably have test coverage for it, but we _could_ say that multiple `playRun` tasks only works when you disable gated rebuild.

File: subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java [hunk 6]
```diff
@@ -33,31 +34,40 @@
...
+        this(fileWatcherFactory, getDefaultQuietPeriod(), isTriggerBuild());
     }
 
     private static long getDefaultQuietPeriod() {
         return Long.getLong(QUIET_PERIOD_SYSPROP, 250L);
     }
 
-    public DefaultFileSystemChangeWaiterFactory(FileWatcherFactory fileWatcherFactory, long quietPeriodMillis) {
+    private static boolean isTriggerBuild() {
+        return Boolean.getBoolean(TRIGGER_BUILD_SYSPROP);
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java]: done

File: subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java (hunk 1 above)
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/internal/filewatch/DefaultFileSystemChangeWaiterFactory.java]: yep, that's much better.

File: subprojects/core/src/main/java/org/gradle/initialization/BuildGateToken.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.initialization;
+
+public interface BuildGateToken {
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/initialization/BuildGateToken.java]: I'm not partial to the "Token" part of the name... Just change it to `BuildGate`?

File: subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/continuous/PlayReloadWaitingIntegrationTest.groovy (hunk 2 above)
Comment (big-guy) [inline subprojects/platform-play/src/integTest/groovy/org/gradle/play/integtest/continuous/PlayReloadWaitingIntegrationTest.groovy]: yep

File: subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildRequestContext.java [hunk 8]
```diff
@@ -20,12 +20,19 @@
 
 public class DefaultBuildRequestContext implements BuildRequestContext {
     private final BuildCancellationToken token;
+    private final BuildGateToken gate;
     private final BuildEventConsumer buildEventConsumer;
     private final BuildRequestMetaData metaData;
 
+    // TODO: Wire the gate token into other levels
```
Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildRequestContext.java]: `BuildRequestContext` is used by all of the entry points (TAPI, launcher, integ tests) when starting a Gradle build.  I need to update each of those and decide if they should be creating a new build gate or reusing one from somewhere else.  Keeping the old constructor let me defer how to handle that. 

This means we _could_ push the build gate up into the TAPI (like the cancellation token) and allow a TAPI client to open/close the gate. I didn't intend on doing that.

Comment (big-guy) [inline subprojects/core/src/main/java/org/gradle/initialization/DefaultBuildRequestContext.java]: I clarified the TODO with some of these details.

File: subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayWorkerClient.java [hunk 9]
```diff
@@ -24,17 +24,21 @@
...
-    private final BuildGateToken gateToken;
+    private final BuildGateToken.GateKeeper gateKeeper;
     private final BlockingQueue<PlayAppStart> startEvent = new SynchronousQueue<PlayAppStart>();
     private final BlockingQueue<PlayAppStop> stopEvent = new SynchronousQueue<PlayAppStop>();
     private int gateCount;
 
     public PlayWorkerClient(BuildGateToken gateToken) {
-        this.gateToken = gateToken;
-        gateToken.addGateKeeper(this);
+        if (Boolean.getBoolean(GATED_BUILD_SYSPROP)) {
```
Comment (bigdaz) [inline subprojects/platform-play/src/main/java/org/gradle/play/internal/run/PlayWorkerClient.java]: Should default to `true` for now. Or perhaps rename the property so setting to 'true' means 'do not wait for browser before building'. `org.gradle.internal.play.eagerRebuild`?

File: subprojects/launcher/src/integTest/groovy/org/gradle/launcher/continuous/ContinuousBuildGateIntegrationTest.groovy [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        succeeds("work")
+        then:
+        outputFile.text == "start"
+
+        when:
+        // Make some file system changes to inputFile
+        inputFile.text = "changed"
+        then:
+        // Gradle has detected the changes to inputFile
+        def pending = server.expectAndBlock("pending")
```
Comment (bigdaz) [inline subprojects/launcher/src/integTest/groovy/org/gradle/launcher/continuous/ContinuousBuildGateIntegrationTest.groovy]: Although unlikely, this is a race condition: the 'pending' call could happen before the `expectAndBlock` is configured.
Instead, it should be:
```
// Set up the block
def pending = server.expectAndBlock('pending')
// Trigger the change that will get us to the block
inputFile.text = 'changed'
// Wait for the block to be reached
pending.waitForAllPendingCalls()
```

Comment (big-guy) [inline subprojects/launcher/src/integTest/groovy/org/gradle/launcher/continuous/ContinuousBuildGateIntegrationTest.groovy]: good catch

**Pull request merged**